src/Microsoft.Health.Fhir.Liquid.Converter/Utilities/ComplexObjectFilterUtility.cs

```diff
@@ -0,0 +1,173 @@
+            }
+
+            // Get our key name
+            var key = path.Dequeue();
+
+            // Check if our key is an object property or an array reference
+            if (key.Contains('['))
+            {
+                // Our required path dictates we should have an array and if not
+                // then this is not a match
+                if (input is object[] inputArray)
+                {
+                    // If key is [] then we want to loop over every object in the array
+                    // and check for our future paths
+                    if (key == "[]")
+                    {
+                        if (path.Count == 0)
+                        {
+                            return (value == null) || inputArray.Contains(value);
+                        }
+
+                        foreach (object obj in inputArray)
+                        {
+                            // Clone the queue to scope the path at this level for each loop
+                            var localPath = new Queue<string>(path);
+                            if (ObjHasValueAtPath(obj, localPath, value))
+                            {
+                                return true;
+                            }
+                        }
+
+                        return false;
+                    }
+
+                    // If our array key references a specific index
+                    else
+                    {
+                        // Trim and parse out our numeric index
+                        var indexStr = key.TrimStart('[').TrimEnd(']');
+                        var index = int.Parse(indexStr);
+
+                        if (index >= 0 && inputArray.Count() > index)
+                        {
+                            // Recurse
+                            return ObjHasValueAtPath(inputArray[index], path, value);
+                        }
+                        else
+                        {
+                            // It's not possible for this array to have the element specified
+                            // so this is not a match
+                            return false;
+                        }
+                    }
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            // Our key is referencing an object property
+            else
+            {
+                if (input is Dictionary<string, object> inputObject)
+                {
+                    if (inputObject.ContainsKey(key))
+                    {
+                        // Recurse
+                        return ObjHasValueAtPath(inputObject[key], path, value);
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        // This function splits a string into an array of parts like
+        // `test.path[].to[5].value` -> ['test', 'path', '[]', 'to', '[5]', 'value']
```

Comment:
Would a path like `test.path[]potato.stuff` result in one of the parts being `[]potato`?  Would this be intentional, or cause unintended effects?

Reply:
Yes it would (result in the array you're describing)

It depends on what you call unintended. That is not a valid path so I would expect it to either return nothing or throw an exception. The code currently throws an exception if the recursion can get to that part of the path.